### PR TITLE
`pack run` to build and immediately run a container from source

### PR DESCRIPTION
**Pull request by scothis**
Opening this PR to collect feedback. It's not yet ready to merge.

Introduces a `pack run` command which builds a container and then
immedatitly runs the newly built container.

A subset of the `pack build` args are supported, including:
- `--path`
- `--builder`
- `--run-image`

And `--port` is introduced as the local port the running container will
receive requests for.

The run command internally leverages the build command's infrastructure.

The reponame is the md5 hash of the working directory.

### CODE REVIEW

**Comment by scothis** [comment]
Known issues:
- [ ] need to add tests
- [ ] `ctl-c` is only impactful after the container starts running, it is ignored prior to that

**Comment by josegonzalez** [comment]
Should this also have a method of pulling in a `.env` file for environment variables? Or perhaps flags for setting env vars?

**Comment by scothis** [comment]
@josegonzalez can you expand on what you would like to see?

**Comment by scothis** [comment]
There are two modifications we should make to this PR before merging:
- [ ] use a common prefix for the built image so that images are identifiable
- [ ] instead of using 8080 inside the container, get the default port from the docker expose metadata for the container

@josegonzalez env vars can be set by individual buildpacks, is that sufficient for you?

**Comment by josegonzalez** [comment]
@scothis not sure what you mean by that? Do you mean a user would have to create an `env-buildpack` to do what I am suggesting here? 

Note: I don't think this is blocking, but more of a usability improvement for those utilizing the tool locally to verify their apps will work as expected on deployment.

File: run.go [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+
+type RunConfig struct {
+	Port  string
+	Build Task
+	// All below are from BuildConfig
+	AppDir   string
+	Builder  string
+	RunImage string
+	RepoName string
+	// Above are copied from BuildFlags are set by init
```
Comment (jchesterpivotal) [inline run.go]: I find the "Above" comments a little confusing, especially with the "All below" comment at the top.

Comment (scothis) [reply run.go]: Would you believe that I copied that comment from BuildConfig, which the first comment is trying to describe? Yea, it's confusing, I'll remove the extra comments

**Comment by scothis** [comment]
@jchesterpivotal ptal

**Comment by jchesterpivotal** [comment]
Thanks. LGTM after change.

File: run.go [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+		Cli:             bc.Cli,
+		Stdout:          bc.Stdout,
+		Stderr:          bc.Stderr,
+		Log:             bc.Log,
+		FS:              bc.FS,
+		Config:          bc.Config,
+		Images:          bc.Images,
+		WorkspaceVolume: bc.WorkspaceVolume,
+		CacheVolume:     bc.CacheVolume,
+	}
```
Comment (ekcasey) [inline run.go]: I not sure why some of these flags are needed on RunConfig. It seems like `AppDir`, `Builder`, `RunImage`,  `Config`, `Images`, `WorkspaceVolume` and `CacheVolume` aren't needed by the implementation of Run. I think they are only needed to generate the BuildConfig, which you are already saving on RunConfig. Let me know if I am missing something...

Comment (scothis) [reply run.go]: You're right, they are not used. I'll remove them.

**Comment by scothis** [comment]
@ekcasey please take another look

**Pull request merged**